mm2src/coins/lightning.rs

```diff
@@ -473,37 +621,512 @@ pub async fn open_channel(ctx: MmArc, req: OpenChannelRequest) -> OpenChannelRes
+    channel_details: ChannelDetailsForRPC,
+}
+
+pub async fn get_channel_details(
+    ctx: MmArc,
+    req: GetChannelDetailsRequest,
+) -> GetChannelDetailsResult<GetChannelDetailsResponse> {
+    let coin = lp_coinfind_or_err(&ctx, &req.coin).await?;
+    let ln_coin = match coin {
+        MmCoinEnum::LightningCoin(c) => c,
+        _ => return MmError::err(GetChannelDetailsError::UnsupportedCoin(coin.ticker().to_string())),
+    };
+    let channel_id =
+        hex::decode(req.channel_id.clone()).map_to_mm(|e| GetChannelDetailsError::DecodeError(format!("{}", e)))?;
+    let channel_details = ln_coin
+        .channel_manager
+        .list_channels()
+        .into_iter()
+        .find(|chan| chan.channel_id == channel_id[..])
+        .ok_or(GetChannelDetailsError::NoSuchChannel(req.channel_id))?
+        .into();
+
+    Ok(GetChannelDetailsResponse { channel_details })
+}
+
+#[derive(Deserialize)]
+pub struct GenerateInvoiceRequest {
+    pub coin: String,
+    pub amount_in_msat: Option<u64>,
+    pub description: String,
+}
+
+#[derive(Serialize)]
+pub struct GenerateInvoiceResponse {
+    invoice: String,
+    payment_hash: String,
+}
+
+/// Generates an invoice (request for payment) that can be paid on the lightning network by another node using send_payment.
+pub async fn generate_invoice(
+    ctx: MmArc,
+    req: GenerateInvoiceRequest,
+) -> GenerateInvoiceResult<GenerateInvoiceResponse> {
+    let coin = lp_coinfind_or_err(&ctx, &req.coin).await?;
+    let ln_coin = match coin {
+        MmCoinEnum::LightningCoin(c) => c,
+        _ => return MmError::err(GenerateInvoiceError::UnsupportedCoin(coin.ticker().to_string())),
+    };
+    let nodes_addresses = ln_coin.nodes_addresses.lock().clone();
+    for (node_pubkey, node_addr) in nodes_addresses {
+        if connect_to_node(node_pubkey, node_addr, ln_coin.peer_manager.clone())
+            .await
+            .is_err()
+        {
+            log::error!(
+                "Channel with node: {} can't be used for invoice routing hints due to connection error.",
+                node_pubkey
+            );
+        }
+    }
+    let network = ln_coin.platform_fields.network.clone().into();
+    let invoice = create_invoice_from_channelmanager(
+        &ln_coin.channel_manager,
+        ln_coin.keys_manager,
+        network,
+        req.amount_in_msat,
+        req.description,
     )?;
+    Ok(GenerateInvoiceResponse {
+        invoice: invoice.to_string(),
+        payment_hash: invoice.payment_hash().to_string(),
+    })
+}
 
-    let mut unsigned_funding_txs = ln_coin.platform_fields.unsigned_funding_txs.lock().await;
-    unsigned_funding_txs.insert(request_id, unsigned);
+#[derive(Clone, Debug, Deserialize, PartialEq)]
+#[serde(tag = "type")]
+pub enum Payment {
+    #[serde(rename = "invoice")]
+    Invoice { invoice: String },
+    #[serde(rename = "keysend")]
+    Keysend {
+        // The recieving node pubkey (node ID)
+        destination: String,
```

Comment:
If it's a pubkey then it's worth using a more specific type here.

Reply:
Done